Pandas
======

Pandas (Python Data Analysis Library)is another a Python library whose purpose is to allow one to do statistical data analysis easily in Python.  Pandas adds a new fundamental data type, the DataFrame to the Python language.  The DataFrame is inspired by the same data type from the R statistical analysis programming envrionment.  DataFrames give us easy ways to load and massage a large data set of values.  The rest of the Pandas library adds methods to do various complex statistical analysis on the data.  Pandas is inspired by, and meant to replace functions
one would do in R with pure Python.  Though a relatively new library, Pandas has quickly matured.  It still does not cover some of the more complex analysis you could do with R, but for many people it will have everything they need already.  Plus it is written in Python to boot, meaning one can avoid the somewhat unintuitive implementation that R provides.  Pandas is also being designed with an eye towards computational performance.  It is already as good, if not often better than R performance wise on very large data sets.

References
----------

* [Pandas web site](http://pandas.pydata.org/)
* [Python for Data Analysis](http://www.amazon.com/Python-Data-Analysis-Wes-McKinney/dp/1449319793/ref=sr_1_1?ie=UTF8&qid=1377800910&sr=8-1&keywords=python+for+data+analysis) by Wes McKinney, 2012. (developer of Pandas) 

In [1]:
import numpy as np
import pandas as pd

Important Features in pandas
============================

Loading Data from delimited text files
--------------------------------------

A simple yet very common data format used by scientists are plain text files, with a delimiter of some kind between each data item.  This basically arranges the data as a table.  By convention, each row of the table represents one observation, or data point.  And each column of the table represents a feature or dimension that was measured or observed.  A common delimiter between each feature is to use a comma ',', though you will see files that use a blank space ' ', a tab '\\t' or sometimes something else to delimit fields.  Pandas allows you to easily load even very large plain text delimited files for analysls.

Here is an example of a comma separated data file (from [Luis Zaman's](http://research.luiszaman.com/) digital parasite data set):

In [2]:
!head data/parasite_data.csv

Virulence,Replicate,ShannonDiversity
0.5,1,0.0592619
0.5,2,1.0936
0.5,3,1.13939
0.5,4,0.547651
0.5,5,0.0659277
0.5,6,1.34433
0.5,7,1.68048
0.5,8,0
0.5,9,2.04768


In [3]:
!tail data/parasite_data.csv

 ,41,0
 ,42,0
 ,43,0
 ,44,0
 ,45,0.391061
 ,46,0.00166884
 ,47,0
 ,48,0.444463
 ,49,0.383512
 ,50,0.511329

Notice that some of the rows at the end have a blank space in the first column.  We will pass pandas a parameter to tell it how to treat these values.

In [4]:
exp_df = pd.read_csv("data/parasite_data.csv", na_values=[" "])

In [5]:
print(exp_df)

     Virulence  Replicate  ShannonDiversity
0          0.5          1          0.059262
1          0.5          2          1.093600
2          0.5          3          1.139390
3          0.5          4          0.547651
4          0.5          5          0.065928
..         ...        ...               ...
345        NaN         46          0.001669
346        NaN         47          0.000000
347        NaN         48          0.444463
348        NaN         49          0.383512
349        NaN         50          0.511329

[350 rows x 3 columns]


Notice that pandas automatically names the column features, using the 1st line from the data file (known as the header).  Notice also that the first feature, Virulance, is missing 50 values (it had 300 non-null values, while the other features had 350 values.

Accessing data in pandas DataFrames
-----------------------------------

We can directly access any column and row by indexing the DataFrame (kinda like a dictionary):

In [6]:
print(exp_df['Virulence'])

0      0.5
1      0.5
2      0.5
3      0.5
4      0.5
      ... 
345    NaN
346    NaN
347    NaN
348    NaN
349    NaN
Name: Virulence, Length: 350, dtype: float64


In [7]:
print(exp_df['ShannonDiversity'][12])

1.58981


We can also access all of the values in a column meeting a certain criteria:

In [8]:
print(exp_df[exp_df['ShannonDiversity'] > 2.0])

     Virulence  Replicate  ShannonDiversity
8          0.5          9           2.04768
89         0.6         40           2.01066
92         0.6         43           2.90081
96         0.6         47           2.02915
105        0.7          6           2.23427
117        0.7         18           2.14296
127        0.7         28           2.23599
129        0.7         30           2.48422
133        0.7         34           2.18506
134        0.7         35           2.42177
139        0.7         40           2.25737
142        0.7         43           2.07258
148        0.7         49           2.38326
151        0.8          2           2.07970
153        0.8          4           2.38474
163        0.8         14           2.03252
165        0.8         16           2.38415
170        0.8         21           2.02297
172        0.8         23           2.13882
173        0.8         24           2.53339
182        0.8         33           2.17865
196        0.8         47       

NumPy use of pandas DataFrame data
----------------------------------

Another powerful feature of the pandas DataFrame is that they are fully integrated with NumPy.  

In [9]:
np.sort(exp_df['ShannonDiversity'])

array([0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00,
       0.00000e+00, 0.00000e+00, 0.00000e+00, 0.00000e+00, 0.000

Blank/Omitted data
==================

One of the toughest preprocessing steps that is needed in analyzing a real data step is trying to identify and figure out how to handle missing data.  Pandas gives great support for identifying and specifying how to handle these missing values.

In [10]:
print(exp_df[np.isnan(exp_df['Virulence'])])

     Virulence  Replicate  ShannonDiversity
300        NaN          1          0.000000
301        NaN          2          0.000000
302        NaN          3          0.833645
303        NaN          4          0.000000
304        NaN          5          0.990309
305        NaN          6          0.000000
306        NaN          7          0.000000
307        NaN          8          0.000000
308        NaN          9          0.061414
309        NaN         10          0.316439
310        NaN         11          0.904773
311        NaN         12          0.884122
312        NaN         13          0.000000
313        NaN         14          0.000000
314        NaN         15          0.000000
315        NaN         16          0.000000
316        NaN         17          0.013495
317        NaN         18          0.882519
318        NaN         19          0.000000
319        NaN         20          0.986830
320        NaN         21          0.000000
321        NaN         22       

DataFrame methods automatically ignore NA/NaN values (which is usually the right thing to do).

In [11]:
print("Mean virulence across all treatments:", exp_df['Virulence'].mean())

Mean virulence across all treatments: 0.7500000000000013


Although we can filter out NA/NaN values if we need to, to for example pass to other libraries that don't handle these so well.  Notice that the following has dropped those 50 rows that had null Virulence values.

In [12]:
print(exp_df.dropna())

     Virulence  Replicate  ShannonDiversity
0          0.5          1          0.059262
1          0.5          2          1.093600
2          0.5          3          1.139390
3          0.5          4          0.547651
4          0.5          5          0.065928
..         ...        ...               ...
295        1.0         46          0.644337
296        1.0         47          0.875816
297        1.0         48          0.000000
298        1.0         49          0.000000
299        1.0         50          0.000000

[300 rows x 3 columns]


Or if needed, we can instead replace all NA/NaN entries with a valid (default) value.

In [13]:
print(exp_df.fillna(0.0)['Virulence'])

0      0.5
1      0.5
2      0.5
3      0.5
4      0.5
      ... 
345    0.0
346    0.0
347    0.0
348    0.0
349    0.0
Name: Virulence, Length: 350, dtype: float64


Some Simple Summary Statistics with Pandas
------------------------------------------

Lets look at some very simple summary statistics with pandas.  Conveniently DataFrames have all kinds of built-in functions to perform standard operations on them en masse: add() sub() mul() div() mean() std() etc.

In [14]:
print("Mean Shannon Diversity w/ 0.8 Parasite Virulence = ", exp_df[exp_df['Virulence'] == 0.8]['ShannonDiversity'].mean())
print("Variance Shannon Diversity w/ 0.8 Parasite Virulence = ", exp_df[exp_df['Virulence'] == 0.8]['ShannonDiversity'].var())

Mean Shannon Diversity w/ 0.8 Parasite Virulence =  1.2691338187999996
Variance Shannon Diversity w/ 0.8 Parasite Virulence =  0.6110384333126732


Some More Examples
------------------

The following are more examples of how you can manipulate pandas DataFrames (taken from the pandas [10-minute tutorial](http://pandas.pydata.org/pandas-docs/stable/10min.html))

In [15]:
dates = pd.date_range('20130101', periods=6)
print(dates)
df = pd.DataFrame(np.random.randn(6,4), index=dates, columns=list('ABCD'))
print(df)

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')
                   A         B         C         D
2013-01-01 -0.361172  1.809067 -0.520126 -0.584595
2013-01-02 -1.275041 -1.014947  0.322499 -0.121134
2013-01-03  0.890945  0.568694  0.228494  0.410192
2013-01-04 -2.070422  0.265733 -0.636263 -0.057367
2013-01-05 -0.518790 -0.548848 -1.373859  0.311020
2013-01-06 -1.192410 -1.744496  0.922961  0.397199


In [16]:
df2 = pd.DataFrame({'A': 1.0,
                    'B': pd.Timestamp('20130102'),
                    'C': pd.Series(1, index=range(4), dtype='float32'),
                    'D': np.array([3] * 4, dtype='int32'),
                    'E': 'foo'})
df2

A          B    C  D    E
0  1.0 2013-01-02  1.0  3  foo
1  1.0 2013-01-02  1.0  3  foo
2  1.0 2013-01-02  1.0  3  foo
3  1.0 2013-01-02  1.0  3  foo

In [17]:
df2.dtypes

A           float64
B    datetime64[ns]
C           float32
D             int32
E            object
dtype: object

In [18]:
df.head()

A         B         C         D
2013-01-01 -0.361172  1.809067 -0.520126 -0.584595
2013-01-02 -1.275041 -1.014947  0.322499 -0.121134
2013-01-03  0.890945  0.568694  0.228494  0.410192
2013-01-04 -2.070422  0.265733 -0.636263 -0.057367
2013-01-05 -0.518790 -0.548848 -1.373859  0.311020

In [19]:
df.tail(3)

A         B         C         D
2013-01-04 -2.070422  0.265733 -0.636263 -0.057367
2013-01-05 -0.518790 -0.548848 -1.373859  0.311020
2013-01-06 -1.192410 -1.744496  0.922961  0.397199

In [20]:
df.columns

Index(['A', 'B', 'C', 'D'], dtype='object')

In [21]:
df.values # notice, the easiest way to turn whole pandas DataFrame into a NumPy array with same rows/columns

array([[-0.36117182,  1.80906666, -0.52012609, -0.58459497],
       [-1.27504093, -1.01494656,  0.32249904, -0.1211336 ],
       [ 0.89094464,  0.56869383,  0.22849447,  0.41019167],
       [-2.07042214,  0.26573254, -0.63626277, -0.05736658],
       [-0.51879024, -0.54884768, -1.37385857,  0.31101963],
       [-1.19240973, -1.74449607,  0.92296145,  0.39719909]])

Describe shows a quick statistics summary of your data:

In [22]:
df.describe()

A         B         C         D
count  6.000000  6.000000  6.000000  6.000000
mean  -0.754482 -0.110800 -0.176049  0.059219
std    1.011151  1.262355  0.822796  0.390239
min   -2.070422 -1.744496 -1.373859 -0.584595
25%   -1.254383 -0.898422 -0.607229 -0.105192
50%   -0.855600 -0.141558 -0.145816  0.126827
75%   -0.400576  0.492954  0.298998  0.375654
max    0.890945  1.809067  0.922961  0.410192

Sort data by values of some column

In [23]:
df.sort_values(by='B')

A         B         C         D
2013-01-06 -1.192410 -1.744496  0.922961  0.397199
2013-01-02 -1.275041 -1.014947  0.322499 -0.121134
2013-01-05 -0.518790 -0.548848 -1.373859  0.311020
2013-01-04 -2.070422  0.265733 -0.636263 -0.057367
2013-01-03  0.890945  0.568694  0.228494  0.410192
2013-01-01 -0.361172  1.809067 -0.520126 -0.584595

Select a single column, which yields a Series, equivalent to df.A

In [24]:
df['A']

2013-01-01   -0.361172
2013-01-02   -1.275041
2013-01-03    0.890945
2013-01-04   -2.070422
2013-01-05   -0.518790
2013-01-06   -1.192410
Freq: D, Name: A, dtype: float64

Copy a column values into a NumPy array, for passing to a library for example:


In [25]:
col_A = df['A'].values
print(col_A)
print(type(col_A))

[-0.36117182 -1.27504093  0.89094464 -2.07042214 -0.51879024 -1.19240973]
<class 'numpy.ndarray'>


Select rows of a DataFrame using slices.

In [26]:
print(df[1:3])
print(df['20130103':'20130105'])

                   A         B         C         D
2013-01-02 -1.275041 -1.014947  0.322499 -0.121134
2013-01-03  0.890945  0.568694  0.228494  0.410192
                   A         B         C         D
2013-01-03  0.890945  0.568694  0.228494  0.410192
2013-01-04 -2.070422  0.265733 -0.636263 -0.057367
2013-01-05 -0.518790 -0.548848 -1.373859  0.311020


Selecting multiple columns.  Unfortunately

In [27]:
df.loc[:,['A','B']]

A         B
2013-01-01 -0.361172  1.809067
2013-01-02 -1.275041 -1.014947
2013-01-03  0.890945  0.568694
2013-01-04 -2.070422  0.265733
2013-01-05 -0.518790 -0.548848
2013-01-06 -1.192410 -1.744496

Getting particular columns and particular rows

In [28]:
df.loc['20130104':'20130106', ['C', 'A']]

C         A
2013-01-04 -0.636263 -2.070422
2013-01-05 -1.373859 -0.518790
2013-01-06  0.922961 -1.192410

Selecting only particular rows by some condition.

In [29]:
df[df.A > 0]

A         B         C         D
2013-01-03  0.890945  0.568694  0.228494  0.410192

# Versions

In [30]:
%reload_ext version_information

%version_information numpy, scipy, matplotlib

Software versions
Python 3.7.4 64bit [GCC 7.3.0]
IPython 7.8.0
OS Linux 5.0.0 27 generic x86_64 with debian buster sid
numpy 1.16.4
scipy 1.3.1
matplotlib 3.1.1
Mon Sep 02 17:27:43 2019 CDT